Close #4 Refactor entrypoint

BraisGabin opened the pull request:
This is my first try to refactorize the entrypoint.

A changes that I thought while I was writing it:
* The default start is --update=mods. If you are running the server you are getting the last version of the game from docker hub, so you only need to update the mods. Possible downside: Can this create more outdated server? I don't think so but...

Feedback?

- [x] Test the configuration file generation

thasmo commented:
Great work so far! I think we should check for game and mod updates on boot, just to provide the best experience possible.

As for the new script itself; some things I think we should consider:
- Have a very simple entrypoint script.
- Move logical, connected script parts to separate files.

This means, it would be nice if we could move code from `build/script/dsta` to the `entrypoint.sh` file, and move all functions ec. to separate files, and either include them with `source filename.sh` or just call them directly.

I think the entrypoint script should hold the general logic i.e. `case $@` etc,
but with very few lines, and delegate detailed commands to dedicated functions in separate files, so that everything is very easy to read and understand.

But in general, very nice work so far!

BraisGabin commented:
Can you elavorete a bit more? With some examples? I agree with you about simplify the entypoint. A separated script that creates the configs.

thasmo commented:
Sure, what I meant was, that the entrypoint script should contain the basic logic in only very few lines:
```sh
# PRE-SCRIPTS
# update game/mods
# create config directory
# generate random server name
# create settings.ini
# create adminlist.txt
# create whitelist.txt
# create blocklist.txt
# handle preset and world generation
# install mods
# configure mods

# CORE-SCRIPTS
# case $@
## start
## update
## log
## help

# FALLBACK
# exec "$@"
```
So it would be easy to see what's going on. All concrete commands i.e. all the pre-scripts and the core scripts, like starting, updating etc. could be moved to separate files e.g. `scripts/command-start.sh` or `scripts/command-log` etc. This means the `dsta` script is splitted up and parts moved to the entrypoint script, and other parts moved to separate files/functions.

Furthermore I think we shouldn't use `dsta start` as `CMD` but rather `start` only and skip the `dsta` command, so usage will be easier and commands in general will be shorter.

BraisGabin commented:
I like your idea: Split the different commands. It improves readability because they do very different things.

But, what do you think about keep `dsta` and create different sub-scripts or *tools*: `dsta-start`, dsta-update` and `dsta-log`. This idea is to parallelise how the `git` shell client works. Git has different tools too: `git-init`, `git-rebase`...

I don't want to merge `dsta` inside the entrypoint because the script can be usefull outside the entrypoint context. Examples:
* `docker exec -it <container> dsta log --chat` in a running instance.
* You can start the instance with a `sleep infinity` and then start the server with the `exec`. Don't know why some one will do this... but we are giving flexibility.

So my proposal:
* Split the code inside `dsta` in three new scripts: `dsta-start`, dsta-update` and `dsta-log`.  `dsta` just selects the correct tool and launchs it.
* Create a new script to generate the configuration.

I just re-read your comment:

> Furthermore I think we shouldn't use `dsta start` as `CMD` but rather `start` only and skip the `dsta` command, so usage will be easier and commands in general will be shorter.

I thought the same. But start and log are shell commands so we can't know which one the user wants to use. Probably ours but... If you check, `postgres` have more less the same problem. I tried to follow more less their path.

thasmo commented:
Thanks for your input, I agree in general.

I think we should keep `dsta` as base command for now, although I'm not sure if we can come up with something better like `dst-server`, `dst`, `server` or something different, because `dsta` means `DST Academy` and doesn't reflect what the Docker image does/is.

What I didn't understand is `Create a new script to generate the configuration.`. What do you mean by that?

BraisGabin commented:
yeah, dsta is not the best name... something starting with w? Something that seems the name of a PJ in the game? It will not reflect what it does but... it's memorable.

And what I was trying to say is: As you say, we must move to other script how we generate the settings. I think that this new one can be just a helper script. No name, and not inside the `PATH`. Just `source settings_builder.sh` and that's it.

thasmo commented:
Ye, but I'd still use multiple scripts, probably.
```sh
source scripts/boot/update.sh
source scripts/boot/settings.sh
source scripts/boot/lists.sh
source scripts/boot/world.sh
source scripts/boot/mods.sh
```
Or something similar.

BraisGabin commented:
It looks good. What about the name? Any idea?

thasmo commented:
I'd vote for `dst-server` although it's a bit long, it reflects the image as best as possible.

BraisGabin commented:
Ok. I made *some* changues. I split all of them in commits to imrpove the readability. What do you think? I recommend you to check the commits and not the whole diff because 22 files with more than 600 new lines is a bit too much.

I must add some tests for the config generation before you can merge this, but I want your input first. And some help because I have no idea what is the lists format (admin, white and block).

BraisGabin commented:
The tests didn't pass for a format error in the help output. I hope Travis agree with me.

thasmo commented:
Great progress!

Bear with me, I got some more feedback and I appreciate your response.

- Get rid of `usage` functions and use `cat usage-<command>.txt` instead.
- Put `if`, `elif` and `then` on the same line: `if [ ... ]; then` for better readability, please.
- Move `server/build/script/dst-server` to the entrypoint script, is that possible?
- Can we use a file-extension `sh` for all files? So we have proper syntax highlighing in IDEs *and* we can get rid of the `.editorconfig` file in `server/build/script`. That would be really nice!
- Add an empty line between `fi` and `if` lines for readability, please.
- I think `if [ `whoami` == "steam" ]` in all the scripts is pretty ugly and makes scripts bloated. Is there any useful way to get rid of it? Do you have an idea?
- For `case` statements, could we add an empty line after `;;` for readability?

So far, I really like where this is heading. Keep up the nice work but don't feel stressed please. :)

BraisGabin commented:
> * Get rid of usage functions and use cat usage-<command>.txt instead. Can we move help text to txt files and just cat them out?

Done.

> * Put if, elif and then on the same line: if [ ... ]; then for better readability, please.
> * Add an empty line between fi and if lines for readability, please.

Done.

> * For case statements, could we add an empty line after ;; for readability?

Hmmm... I think that this is too much empty lines. The `;;` line is nearly an empty line and the next one has a different indention. That seems enought for me. Do you really think that it improves readability? I don't mind to add extra empty lines to improve readability, but the scroll size it's important too.

> * I think if [whoami== "steam" ] in all the scripts is pretty ugly and makes scripts bloated. Is there any useful way to get rid of it? Do you have an idea?

I didn't like too much neigher. I didn't follow the "Don't repeat yourself" pattern. I improved the code a bit. What do you think now?

> * Move server/build/script/dst-server to the entrypoint script, is that possible?

I think that the people can use the image in different ways. If we have `dst-server` script (I don't like the name... but I don't find any better...) the users can run the image with `docker run dstacademy/server bash` do whatever they want and finally run the server with `dst-server start`. I use it in development. I don't know if someone'll use it in production but I think that it didn't hurt.

From the Docker documentation:

> An `ENTRYPOINT` allows you to configure a container that will run as an executable.

Right that's what our `ENTRYPOINT` do. Just configure the container and let other script to do the run job. There are a lot of example in the official images with this behavior.

> * Can we use a file-extension sh for all files? So we have proper syntax highlighing in IDEs and we can get rid of the .editorconfig file in server/build/script. That would be really nice!

I think is better how it's right now for the reason I give in the previus point. It's easier, cleaner and and nicer to execute `steamcmd` than `steamcmd.sh`.

The highlighing is a problem... it works for me in atom.

---

If you are not conviced about some of this points please reply me.

I edited the first post with a TODO list.

BraisGabin commented:
I Added the tests and, of course, I fixed the bugs detected with them.

I didn't test `boot/settings.sh` I want to recreate that file from scratch. But not in `0.4`.

thasmo commented on build/docker-entrypoint.sh:
```diff
@@ -1,143 +1,21 @@
-CONF_DIR="config"
-USER="steam"
+if [ $1 == "dst-server" ]
+then
```
Please move `then` to line 3: `if [ $1 == "dst-server" ]; then`.

thasmo commented on build/docker-entrypoint.sh:
```diff
@@ -1,143 +1,21 @@
-  echo "$MODS_OVERRIDES" > $FILE_MODS_OVERRIDES
-elif [ -n "$MODS" ] && [ ! -f $FILE_MODS_OVERRIDES ]; then
-  echo "return {" >> $FILE_MODS_OVERRIDES
+  #Configure the server
```
Please add a space. :) `#[SPACE]Configure ...`.

thasmo commented on build/Dockerfile:
```diff
@@ -6,36 +6,38 @@ RUN dpkg --add-architecture i386 \
 	&& apt-get clean \
-	&& rm -rf /var/lib/apt/lists/*
-
-# Install gosu.
```
I split those into two layers for caching optimization. These should be two image layers, not one. There's no gain in combining those two into one layer, is there?

BraisGabin replied:
Are we going to change the `gosu` more thant the `apt-get` part? I don't think so. I think that if we are going to change something it'll be add or remove some `apt-get`. I don't see any cache optimization there. Any way, if you prefer to split it is ok for me. It's not a big deal.

thasmo replied:
Can't follow you. :) The `RUN` directive represents/creates one layer. The layer gets cached for the build process if it succeeds. If we have two layers, one for `apt-get` and one for `gosu`, we can change the `gosu` layer and the `apt-get` layer will still be cached. If we only have one layer, and we change *anything* in that layer, the whole layer has to be rebuilt. That means, having two layers, is more flexible. I don't see any advantage using one layer, in this case.

Am I missing something?

BraisGabin replied:
What I'm arguing is: are we going to change the `gosu` layer? I think that we are going to change more the `atp-get` layer. Any way have one or two layers here is not a big deal. I'll remove that commit.

thasmo commented on build/Dockerfile:
```diff
@@ -6,36 +6,38 @@ RUN dpkg --add-architecture i386 \
 
+# We just copy the necessary scripts to download the game. This improves the development speed
+# because we don't need to download the game again if it's not necessary.
+COPY /script/steamcmd /script/update.sh /usr/local/bin/
```
Obsolete `COPY` line?

BraisGabin replied:
No, we need both `COPY`. The frist one copys the required files to download the game. The second one the rest ones. I don't merge them for cache reasons.

thasmo replied:
Can we copy all files here and remove the line below, no?

thasmo replied:
I mean, maybe I don't understand it. What cache reasons are you talking about?

BraisGabin replied:
The files inside `/script/` change *a lot*. The game change less than this files. At list now that we are in development. And download the game takes *a lot* of time. If we copy the files after we download the game we don't need to download the game again. I can remove the first `COPY` but I'll need to use the `steamcmd.sh` command.

Any way I think that when we release `v1.0` we can merge both in the top. In that momment (I hope) the game'll change more thant those files.

thasmo replied:
Got you. That's true and I'd love to find a way to cache the Steam download some way for development. For now we can keep it that way.

thasmo commented on build/Dockerfile:
```diff
@@ -6,36 +6,38 @@ RUN dpkg --add-architecture i386 \
 	&& update.sh
 
+# Install all the needed scripts
+COPY /script/* /usr/local/bin/
```
Obsolete `COPY` line?

thasmo commented on build/Dockerfile:
```diff
@@ -72,5 +74,6 @@ RUN chown -R steam:steam /usr/local/lib/dsta/ \
 VOLUME ["/var/lib/dsta/config"]
 
-WORKDIR /opt/dst/bin/
+WORKDIR /var/lib/dsta/config
```
Why did the `WORKDIR` change?

BraisGabin replied:
We don't need to be in the /opt/dst/bin/ to run the server now. `dedicated_server_nullrenderer` script do the trick. I think that the more interesant directory in all the container is that. We can even remove the `WORKDIR` but I think that it's good because it help to discovery the dir.

thasmo replied:
If we don't it and as we've already got the path in the `VOLUME`, we should remove it, I think.

thasmo commented on build/script/steamcmd:
```diff
@@ -1,8 +1,7 @@
+  gosu="gosu steam"
 fi
+
+$gosu /opt/steam/steamcmd.sh "$@"
```
Haha, clever. :)

thasmo commented:
So far the PR looks good to me. If you aren't going to add anything more to the PR, I'll merge it into the develop branch soon.
